src/api/v4/layer/layer.js

```diff
@@ -1,90 +1,274 @@
+ * 
+ * @return {carto.source.Dataset|carto.source.SQL} Current source.
+ *
+ * @api
+ */
+Layer.prototype.getSource = function () {
+  return this._source;
+};
+
+/**
+ * Set new columns for `featureClick` events
+ * 
+ * @param {Array<string>} columns An array containing column names.
+ * @return {carto.layer.Layer} this
+ *
+ * @api
+ */
+Layer.prototype.setFeatureClickColumns = function (columns) {
+  this._featureClickColumns = columns;
+  if (this._internalModel) {
+    this._internalModel.infowindow.update(_getInteractivityFields(columns));
+  }
+
+  return this;
+};
+
+/**
+ * Return the columns available in `featureClicked` events.
+ * 
+ * @return  {[Array<string>]} Column names available in `featureClicked` events
+ *
+ * @api
+ */
+Layer.prototype.getFeatureClickColumns = function (columns) {
+  return this._featureClickColumns;
+};
+
+/**
+ * Set new columns for `featureOver` events
+ * 
+ * @param {Array<string>} columns An array containing column names
+ * @return {carto.layer.Layer} this
+ *
+ * @api
+ */
+Layer.prototype.setFeatureOverColumns = function (columns) {
+  this._featureOverColumns = columns;
+  if (this._internalModel) {
+    this._internalModel.tooltip.update(_getInteractivityFields(columns));
   }
-  return this._internalModel._engine.reload();
+
+  return this;
+};
+
+/**
+ * Return the columns available in `featureOver` events.
+ * 
+ * @return  {[Array<string>]} Column names available in `featureOver` events
+ *
+ * @api
+ */
+Layer.prototype.getFeatureOverColumns = function (columns) {
+  return this._featureOverColumns;
 };
 
+/**
+ * Hides the layer.
+ *
+ * @api
+ */
 Layer.prototype.hide = function () {
   this._visible = false;
   if (this._internalModel) {
-    this._internalModel.hide();
+    this._internalModel.set('visible', false);
   }
+
+  return this;
 };
 
+/**
+ * Shows the layer.
+ *
+ * @api
+ */
 Layer.prototype.show = function () {
   this._visible = true;
   if (this._internalModel) {
-    this._internalModel.show();
+    this._internalModel.set('visible', true);
   }
+
+  return this;
 };
 
-Layer.prototype.setSource = function (source) {
-  this._source = source;
-  if (this._internalModel) {
-    this._internalModel.set('source', source);
+// Private functions.
+
+Layer.prototype._createInternalModel = function (engine) {
+  return new CartoDBLayer({
+    id: this._id,
+    source: this._source.$getInternalModel(),
+    cartocss: this._style.toCartoCSS(),
+    visible: this._visible,
+    infowindow: _getInteractivityFields(this._featureClickColumns),
+    tooltip: _getInteractivityFields(this._featureOverColumns)
+  }, { engine: engine });
+};
+
+Layer.prototype._reloadEngine = function () {
+  if (this._engine) {
+    return this._engine.reload();
   }
+
+  return Promise.resolve();
 };
 
+// Internal functions.
+
 Layer.prototype.$setEngine = function (engine) {
+  if (this._engine) {
+    return;
+  }
+  this._engine = engine;
   this._source.$setEngine(engine);
```

Comment:
What happens if `source` has already an engine and it's different from this one?

Reply:
Great question!

It won't set the engine on `this_source`. (https://github.com/CartoDB/cartodb.js/blob/b23453f0620c48b54fbd3a195648329dbca99af0/src/api/v4/source/base.js#L39-L41) because it was already set. Side-effects of this are probably not very good 😄 .

We should throw an error in that case and `src/api/v4/client` should handle it properly (layer shouldn't be added/created).

Also, I think we're not handling this gracefully for dataviews either (see https://github.com/CartoDB/cartodb.js/blob/public-api/src/api/v4/dataview/base.js#L221-L227).

Reply:
I guess it's related to https://github.com/CartoDB/cartodb.js/issues/1868, right @IagoLast?

Reply:
@ivanmalagon 

```
IF the Layer has not been instantiated:
  - Add the source to the layer
IF the layer has been instantiated:
  IF the source has no engine:
    - Add the source to the layer
  IF the source has an engine:
    IF source.engine === layer.engine
      - Add the source to the layer
    IF source.engine !== layer.engine
      - Throw an error. // This is your case
```

Reply:
yep, is a new case.

The layer has not been instantiated (has no engine/client) but the source has an engine.

When the layer is added to a client.
```
IF client.engine === source.engine : 
  (all ok)
ELSE: 
  Error('A source cant be shared between clients'
```

Reply:
We'll review this in the above mentioned issue then.